sqlglot/dialects.py

```diff
@@ -206,33 +206,37 @@ def ts_or_ds_to_date(self, expression):
     }
 
     functions = {
-        'APPROX_COUNT_DISTINCT': lambda args: exp.ApproxDistinct(this=args[0]),
-        'COLLECT_LIST': lambda args: exp.ArrayAgg(this=args[0]),
-        'DATE_ADD': lambda args: exp.DateAdd(this=exp.DateStrToDate(this=args[0]), expression=args[1]),
-        'DATEDIFF': lambda args: exp.DateDiff(
-            this=exp.DateStrToDate(this=args[0]),
-            expression=exp.DateStrToDate(this=args[1]),
-        ),
-        'DATE_SUB': lambda args: exp.DateAdd(
-            this=exp.DateStrToDate(this=args[0]),
-            expression=exp.Star(this=args[1], expression=Token.number(-1)),
-        ),
-        'DATE_FORMAT': lambda args: exp.TimeToStr(this=args[0], format=args[1]),
-        'DAY': lambda args: exp.Day(this=exp.TsOrDsToDate(this=args[0])),
-        'FROM_UNIXTIME': lambda args: exp.UnixToStr(
-            this=args[0],
-            format=list_get(args, 1) or Hive.TIME_FORMAT,
-        ),
-        'GET_JSON_OBJECT': lambda args: exp.JSONPath(this=args[0], path=args[1]),
-        'LOCATE': lambda args: exp.StrPosition(this=args[1], substr=args[0], position=list_get(args, 2)),
-        'MONTH': lambda args: exp.Month(this=exp.TsOrDsToDate(this=args[0])),
-        'PERCENTILE': lambda args: exp.Quantile(this=args[0], quantile=args[1]),
-        'SIZE': lambda args: exp.ArraySize(this=args[0]),
-        'TO_DATE': lambda args: exp.TsOrDsToDateStr(this=args[0]),
-        'UNIX_TIMESTAMP': lambda args: exp.StrToUnix(
-            this=args[0],
-            format=list_get(args, 1) or Hive.TIME_FORMAT,
-        ),
+        'APPROX_COUNT_DISTINCT': exp.ApproxDistinct.from_arg_list,
+        'COLLECT_LIST': exp.ArrayAgg.from_arg_list,
+        'DATE_ADD': lambda args: exp.DateAdd.from_arg_list([exp.DateStrToDate(this=list_get(args, 0)), *args[1:]]),
+        'DATEDIFF': lambda args: exp.DateDiff.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
+            exp.DateStrToDate(this=list_get(args, 1)),
+            *args[2:]
+        ]),
+        'DATE_SUB': lambda args: exp.DateAdd.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
+            exp.Star(this=list_get(args, 1), expression=Token.number(-1)),
+            *args[2:]
+        ]),
+        'DATE_FORMAT': exp.TimeToStr.from_arg_list,
+        'DAY': lambda args: exp.Day.from_arg_list([exp.TsOrDsToDate(this=list_get(args, 0)), *args[1:]]),
+        'FROM_UNIXTIME': lambda args: exp.UnixToStr.from_arg_list([
+            list_get(args, 0),
+            list_get(args, 1) or Hive.TIME_FORMAT,
+            *args[2:]
+        ]),
+        'GET_JSON_OBJECT': lambda args: exp.JSONPath.from_arg_list([list_get(args, 0), list_get(args, 1), *args[3:]]),  # Ignore the 3rd argument
```

Comment:
what do this mean? ignore the 3rd argument, why do you even have to do anything?

Reply:
Check out the following test case - [link](https://github.com/tobymao/sqlglot/blob/main/tests/test_dialects.py#L340-L345). The 3rd argument is indeed ignored there. The reason we pass `, *args[3:]` is to make sure there are no excessive arguments beyond the 3rd one.

Reply:
ok, i think i found a cleaner solution, let me know what you think

```diff --git a/sqlglot/parser.py b/sqlglot/parser.py
index e002c92..f6af8de 100644
--- a/sqlglot/parser.py
+++ b/sqlglot/parser.py
@@ -802,7 +802,9 @@ class Parser:
             elif not callable(function):
                 this = exp.Anonymous(this=this, expressions=args)
             else:
-                this = function(args)
+                this = function(*args)
+                if len(args) > len(this.arg_types):
+                    self.raise_error('too many args')
```

two possible changes

first change, splat out args, and then when you define lambdas, you can do like
lambda x, y, *args: ... 
instead of having to do this *args[3:], really don't like that, you can also name the arguments coming in so it's more readable

also, in here, we can easily raise exception when extranous args are passed notice i can trivially check the initial arg length and compare it with the return type arg length

Reply:
obviously this example code doesn't take into account varargs, so it breaks on arrays

Reply:
I think I like the idea of moving the error handling, but not sure about the lambda x, y, *args: .... If there are not enough arguments the error will still be confusing:
```python
>>> f = lambda a, b, *args: print(a)
>>> f(*[1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: <lambda>() missing 1 required positional argument: 'b'
```

Reply:
makes sense

Reply:
So it seems like we can't really push the validation into this layer. For some dialects we actually want to ignore some arguments while constructing the tree. If we unconditionally fail in the parser we won't be able to support use cases like:
```
DATE_ADD('day', 1, x)
```
for Presto dialect for which we actually ignore the first `'day'` argument. 
@tobymao seems like the validation should still take place in the lambda and subsequently in `from_arg_list`. What do you think?

Reply:
Fixed as per our conversation.